### PR TITLE
WIP: Refactor content type parser

### PR DESCRIPTION
**Pull request by delvedor**
Hi all!
With this pr we change how works our internal body parsers, some users asked us if we could provide a better content type parser API, not everybody wants to handle a stream and we are already doing that internally (with a very low overhead). This pr updates the `addContentTypeParser` API by adding an options object, if the user passes  `{ asString: true }` it will get the body as string, with `{ asBuffer: true }` it will get the body as buffer.

To avoid breaking the compatibility with the ecosystem if the option object is not provided the callback function will be the same as before `function (rawRequest, done)`, instead, if the options object is provided (with the above options) the function signature will be  `function (request, reply, body, done)`.

The only breaking change we are introducing in the API, is the rename of `jsonBodyLimit` option, that becomes `bodyLimit`.

From now it will be possible overwrite our default JSON parser.
I'll update the documentation once we all agree on the API.

There are no significant changes in the benchmarks 🎉 

Related: #643 #707 #782

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: fastify.js [hunk 1]
```diff
@@ -114,9 +114,9 @@ function build (options) {
     server.on('clientError', handleClientError)
   }
 
-  // JSON body limit option
-  validateBodyLimitOption(options.jsonBodyLimit)
-  fastify._jsonBodyLimit = options.jsonBodyLimit || DEFAULT_JSON_BODY_LIMIT
+  // body limit option
+  validateBodyLimitOption(options.bodyLimit)
+  fastify._bodyLimit = options.bodyLimit || DEFAULT_BODY_LIMIT
```
Comment (allevo) [inline fastify.js]: Should we allow 0 to indicate "no limit" ?

File: lib/ContentTypeParser.js [hunk 2]
```diff
@@ -1,71 +1,184 @@
...
+
+function rawBody (request, reply, options, parser, handler) {
+  var asString = parser.asString
+  var limit = options.limit
+  var contentLength = request.headers['content-length'] === undefined
+    ? NaN
+    : Number.parseInt(request.headers['content-length'], 10)
+
+  if (contentLength > limit) {
+    reply.code(413).send(new Error('Request body is too large'))
```
Comment (jsumners) [inline lib/ContentTypeParser.js]: Since the `Error` is duplicated in the `onData` handler, I think it would be wise to use a variable for it. Thus, if the message needs to be updated for some reason, it only has to be updated in one place.

**Comment by jsumners** [comment]
Why `function (req, reply, body, done)` instead of `function (err, body, done)` or just `function (body, done)`?

**Comment by delvedor** [comment]
> Why function (req, reply, body, done) instead of function (err, body, done) or just function (body, done)?

I thought about that, but I'd like to make this api future proof. It can happen that a parsers needs the request headers or set the response headers, and if we pass only `body, done` it will not be possible.

File: lib/ContentTypeParser.js (hunk 2 above)
Comment (delvedor) [reply lib/ContentTypeParser.js]: Correct me if I'm wrong, but if we cache the error object the stack trace will not be super easy to read.

Comment (jsumners) [reply lib/ContentTypeParser.js]: I'm not sure. Could just cache the string.

**Comment by jsumners** [comment]
Then I think the documentation will need some recommendations about what should and should not be done with the `req/reply` within the callback. Also, it should be documented that `done(null, parsedPayload)` will make the parsed payload available as `req.body` within the route handler.

**Comment by delvedor** [comment]
> Then I think the documentation will need some recommendations about what should and should not be done with the req/reply within the callback. Also, it should be documented that done(null, parsedPayload) will make the parsed payload available as req.body within the route handler.

Absolutely.

File: lib/ContentTypeParser.js [hunk 3]
```diff
@@ -1,71 +1,184 @@
...
 
-ContentTypeParser.prototype.getHandler = function (contentType) {
-  if (contentType === undefined) return this.customParsers['']
-
+ContentTypeParser.prototype.getParser = function (contentType) {
   for (var i = 0; i < this.parserList.length; i++) {
     if (contentType.indexOf(this.parserList[i]) > -1) {
-      return this.customParsers[this.parserList[i]]
+      var parser = this.customParsers[this.parserList[i]]
+      this.cachedParsers[contentType] = parser
```
Comment (allevo) [inline lib/ContentTypeParser.js]: This can be a security issue

File: fastify.js (hunk 1 above)
Comment (delvedor) [reply fastify.js]: Allow no limit means create the security issue. I don't agree.
The user can configure it with a big value anyway.

File: lib/ContentTypeParser.js (hunk 3 above)
Comment (delvedor) [reply lib/ContentTypeParser.js]: Why?

File: fastify.js (hunk 1 above)
Comment (jsumners) [reply fastify.js]: That should be up to the user. But I would choose `-1` to remove the limit. It could be that someone would want to use `0` to restrict receiving post bodies altogether.

Comment (allevo) [reply fastify.js]: -1 fine for me.

File: lib/ContentTypeParser.js (hunk 3 above)
Comment (allevo) [reply lib/ContentTypeParser.js]: If I use `"application/json" + randomString()` as "content-type" header, that object grows a lot.

File: fastify.js (hunk 1 above)
Comment (delvedor) [reply fastify.js]: Why would a user do this?

Comment (allevo) [reply fastify.js]: No sorry, my badf.

If you need to parse a *very* large body you should use stream with a custom parser. Let's keep it.

File: lib/ContentTypeParser.js (hunk 3 above)
Comment (allevo) [reply lib/ContentTypeParser.js]: Great!

File: fastify.js (hunk 1 above)
Comment (nwoltman) [reply fastify.js]: Using `Number.MAX_SAFE_INTEGER` would set the limit to about `8 PB`. I'd say that's pretty close to having no upper limit.

File: lib/ContentTypeParser.js [hunk 4]
```diff
@@ -1,67 +1,181 @@
...
-    this.customParsers[contentType] = fn
+    if (contentType !== 'application/json') {
+      this.parserList.unshift(contentType)
+    }
+    this.customParsers[contentType] = parser
   }
 }
 
 ContentTypeParser.prototype.hasParser = function (contentType) {
+  if (contentType === 'application/json') return false
```
Comment (nwoltman) [inline lib/ContentTypeParser.js]: What about after a user sets a custom JSON parser? I think this would handle that case:

```js
if (contentType === 'application/json') {
  return this.customParser['application/json'] !== defaultJsonParser
}
```

File: lib/ContentTypeParser.js [hunk 5]
```diff
@@ -1,67 +1,181 @@
...
+  if (parser === undefined) {
+    reply.code(415).send(new Error('Unsupported Media Type: ' + contentType))
+  } else {
+    if (parser.asString === true || parser.asBuffer === true) {
+      rawBody(
+        request,
+        reply,
+        reply.context._parserOptions,
+        parser,
+        handler
```
Comment (nwoltman) [inline lib/ContentTypeParser.js]: I think you can pass `done` here instead of `handler`. That way the `done` function won't need to be duplicated (line 159).

File: lib/ContentTypeParser.js [hunk 6]
```diff
@@ -1,67 +1,181 @@
...
     }
+  }
+}
 
-    request.body = body
-    handler(reply)
+function defaultJsonParser (request, reply, body, done) {
+  try {
+    var json = JSON.parse(body)
+    done(null, json)
```
Comment (nwoltman) [inline lib/ContentTypeParser.js]: This should be called outside of the try-catch (so that you don't catch errors in `done` and then call the `done` function a second time).

File: lib/ContentTypeParser.js [hunk 7]
```diff
@@ -1,67 +1,174 @@
...
   if (contentType.length === 0) throw new Error('The content type cannot be an empty string')
-  if (typeof fn !== 'function') throw new Error('The content type handler should be a function')
+  if (typeof parser !== 'function') throw new Error('The content type handler should be a function')
 
   if (this.hasParser(contentType)) {
     throw new Error(`Content type parser '${contentType}' already present.`)
   }
+
+  parser.asString = opts.asString === true
+  parser.asBuffer = opts.asBuffer === true
```
Comment (mcollina) [inline lib/ContentTypeParser.js]: what is that `parser` object? Is it a function? I would prefer if we didn't attach properties to it, but we rather wrap it in its own object with these properties.

Also, I would prefer we had a single option in the opts instead so we don't have a problem to validate these.

File: lib/ContentTypeParser.js [hunk 8]
```diff
@@ -1,67 +1,174 @@
+/* eslint-disable no-useless-return */
 'use strict'
 
+const lru = require('tiny-lru')
```
Comment (mcollina) [inline lib/ContentTypeParser.js]: why is the cache need at all?

Comment (delvedor) [reply lib/ContentTypeParser.js]: To skip the `getParser` function.
https://github.com/fastify/fastify/blob/525d68a815983a94fbece88fd4d158a10e17edeb/lib/ContentTypeParser.js#L50-L60
Often it could happen that we get request with the charset among the content type, to handle that case we are using `indexOf`. By using the cache object we just skip the for loop and the index of check.
Why a lru cache and not a normal object? As @allevo pointed out a basic object could lead to a security issues, because an attacker can send dozens of request with slightly different content type.

File: lib/ContentTypeParser.js (hunk 7 above)
Comment (delvedor) [reply lib/ContentTypeParser.js]: > what is that parser object? Is it a function? 

A function

> I would prefer if we didn't attach properties to it, but we rather wrap it in its own object with these properties.

Ok, we can do that

> Also, I would prefer we had a single option in the opts instead so we don't have a problem to validate these.

I'm not sure I have understood, can you rephrase, maybe with an example?

File: lib/ContentTypeParser.js (hunk 8 above)
Comment (mcollina) [reply lib/ContentTypeParser.js]: 👍

File: lib/ContentTypeParser.js (hunk 7 above)
Comment (mcollina) [reply lib/ContentTypeParser.js]:
```js
{ parseAs: 'string' }
```

Which translated to boolean flags. But in this way it's clear to the user only a single flag is possible.

Comment (delvedor) [reply lib/ContentTypeParser.js]: Done

**Comment by mcollina** [comment]
On second thought, I'm not super fond on passing through the reply object. What would happen if somebody called `reply.send()` in there?

Why don't we keep the same API as before? Doing anything with `request` and `reply` has not come up until know, and it was done here only as a precaution.

**Comment by delvedor** [comment]
@mcollina what do you mean by "same api"?
`(body, done)`? `(req, body, done)`?
As I said [here](https://github.com/fastify/fastify/pull/785#issuecomment-365619728) I used this signature to be more future proof. In any case if we want to remove reply, we should keep the request object (at least the raw node request) in the case of the user needs to access the headers.

**Comment by mcollina** [comment]
That can work, yes.

But you can pass the `body` argument only if `parseAs: 'string'` (o `'buffer'`), because otherwise you would have `parser(req, req, fn)`.

File: test/custom-parser.test.js [hunk 9]
```diff
@@ -450,3 +450,183 @@ test('cannot add custom parser after binding', t => {
...
+  t.plan(6)
+  const fastify = Fastify()
+
+  fastify.post('/', (req, reply) => {
+    reply.send(req.body)
+  })
+
+  fastify.addContentTypeParser('application/json', { parseAs: 'buffer' }, function (req, body, done) {
+    t.ok('called')
+    t.ok(Buffer.isBuffer(body))
```
Comment (mcollina) [inline test/custom-parser.test.js]: you should use  `t.type` or `body instanceof Buffer` .

**Comment by delvedor** [comment]
@fastify/fastify I've updated the documentation, please check it out :)

File: docs/ContentTypeParser.md [hunk 10]
```diff
@@ -24,14 +24,29 @@ You can also use the `hasContentTypeParser` API to find if a specific content ty
...
 if (!fastify.hasContentTypeParser('application/jsoff')){
   fastify.addContentTypeParser('application/jsoff', function (req, done) {
-    //code to parse request body /payload for given content type 
+    //code to parse request body /payload for given content type
   })
 }
 ```
 
+#### Body Parser
+You can parse the body of the request in two ways, the first one is showed above, you just add a custom content type parser and handle the request stream. In the second one you should pass a `parseAs` option to the `addContentTypeParser` API, where you declare how you want to get the body, it could be `'string'` or `'buffer'`. If you use the `parseAs` option Fastify will internally handle the stream and perform some checks, such as the [maximum size](https://github.com/fastify/fastify/blob/master/docs/Factory.md#factory-body-limit) of the body and the content length.
```
Comment (jsumners) [inline docs/ContentTypeParser.md]: "You can parse the body of the request in two ways, the first one is showed above, you just add a custom content type parser and handle the request stream." => "You can parse the body of the request in two ways. The first one is shown above: you add a custom content type parser and handle the request stream."

File: docs/ContentTypeParser.md [hunk 11]
```diff
@@ -1,11 +1,11 @@
 <h1 align="center">Fastify</h1>
 
 ## Content Type Parser
```
Comment (jsumners) [inline docs/ContentTypeParser.md]: There should be some information in this document about how the parsed content is made available in the request handler. It is not obvious from the documentation that the handler should expect the parsed content to be `request.body`.

**Comment by mcollina** [comment]
still LGTM
